src/kernel/pagecache.c

```diff
@@ -1333,11 +1349,18 @@ u64 pagecache_get_node_length(pagecache_node pn)
     return pn->length;
 }
 
-closure_function(0, 1, boolean, pagecache_page_release,
+closure_function(1, 1, boolean, pagecache_page_release,
+                 pagecache, pc,
                  rbnode, n)
 {
     pagecache_page pp = struct_from_list(n, pagecache_page, rbnode);
-    refcount_release(&pp->refcount);
+    if (!pp->evicted)
+        refcount_release(&pp->refcount);
+    pagecache pc = bound(pc);
+    pagecache_lock_state(pc);
+    pagelist_remove(&pc->free, pp);
```

Comment:
Can we assert that pp is actually in the free state here? Could something else be possibly holding a reference to it at this point? (e.g. a dirty page has not completed writing)

Perhaps it would be safer to flag the page to be terminated (instead of being added to the free list) and move the pagecache_page deallocate to pagecache_page_free().

Reply:
> Moving the pagecache_page deallocate to pagecache_page_free() cannot be done with the current code, because if pagecache_page_free() is called as a result of a page eviction, the page is still referenced by its node

Right, that's what I meant by flagging the page - to set a boolean within the page indicating that it should not be added to the free list but rather deallocated altogether. Though I guess that doesn't catch pages for the node that have already been evicted and are sitting on the free list.

If, by definition, a pagecache node being released means no outstanding page references are possible, I think it's fine as-is (with an assert). Maybe a comment could elucidate this.

Reply:
Adding an assert that pp is in the free state here revealed a missing refcount release after a write, and upon further inspection the code that handles refcounting was shown to be racy with regards to the pagecache drain functionality. I tried to fix these issues in a separate PR (#1754), and the changes to pagecache_page_release() are now in a commit in that PR.